### ISSUE TITLE
How can I get first day and last day from current calendar whether it is in weekly,monthy or 2 weeks preview?

### ISSUE THREAD

**Issue by CodingWithTashi** [aleksanderwozniak/table_calendar#516]
I was wondering how can I get start date and end date from preview( Weekly,Monthly or two weeks)? Other than that All good.

**Comment by aradhwan** [comment]
I have the same issue.
In previous versions up to 2.3.3 I can get it from:
`onVisibleDaysChanged: (first, last, calendarFormat) { }`
But this is no longer available in version 3.0.0

**Comment by aleksanderwozniak** [comment]
This is a duplicate of #441

**Comment by CodingWithTashi** [comment]
> I was wondering how can I get start date and end date from preview( Weekly,Monthly or two weeks)? Other than that All good.



> This is a duplicate of #441

It doest make sense, This solution work only with month preview. for week and two week preview, first date and last date not able to calculate

**Comment by alvindrakes** [comment]
> I have the same issue.
> In previous versions up to 2.3.3 I can get it from:
> `onVisibleDaysChanged: (first, last, calendarFormat) { }`
> But this is no longer available in version 3.0.0

Hi everyone, may I know what's the equivalent value of `first`, `last` from package v2.3.3 in the current null safe version? 

How do I get the same value from `focusedDay` of `onPageChanged`?

**Comment by brian-quah** [comment]
> It doest make sense, This solution work only with month preview. for week and two week preview, first date and last date not able to calculate

>Hi everyone, may I know what's the equivalent value of first, last from package v2.3.3 in the current null safe version?
>
>How do I get the same value from focusedDay of onPageChanged?

In case anyone still needs a solution for this, you can do something like this:

In your screen State, you declare the instance variable for the calendar format:
```dart
CalendarFormat _calendarFormat = CalendarFormat.month; // or whatever initial format you want
```
Then when creating instantiating the calendar widget, include `onFormatChanged` and `onPageChanged` as follows:
```dart
TableCalendar(
...
    onFormatChanged: (format) {
      if (_calendarFormat != format) {
        setState(() => _calendarFormat = format);
      }
    },
    onPageChanged: (focus) {
      DateTime start, end;
      if (_calendarFormat == CalendarFormat.week) {
        start = DateTime.utc(focus.year, focus.month, focus.day - focus.weekday + 1);
        end = DateTime.utc(focus.year, focus.month, focus.day + 7 - focus.weekday);
      } else if (_calendarFormat == CalendarFormat.twoWeeks) {
        start = DateTime.utc(focus.year, focus.month, focus.day - focus.weekday + 1);
        end = DateTime.utc(focus.year, focus.month, focus.day + 14 - focus.weekday);
      } else { // _calendarFormat == CalendarFormat.month
        start = DateTime.utc(focus.year, focus.month);
        end = DateTime.utc(focus.year, focus.month + 1, 0);
      }
      // TODO: whatever you need to do with the start and end dates
    }
...
}
```

**Comment by mjarpitanand** [comment]
@brian-quah I was looking for exact this solution, this seems to work as somooth as butter. Can you explain me a little bit For two weeks scenario - start day is calculated by ```focusday.day - focusday.weekday + 1 ```

**Comment by mjarpitanand** [comment]
@brian-quah  This is a wrong implemenation above one wont work for two weeks.

**Comment by brian-quah** [comment]
@mjarpitanand 
Sorry for the late reply.
Good catch, after some debugging, seems the date that is chosen as the 'focus' isn't consistent as I had assumed - seems it doesn't always end up focused on the same date despite navigating to the same two-week period from other parts of the calendar.

Looks like we unfortunately have no convenient way of deriving the start and end dates because of this issue.
@aleksanderwozniak Can this issue be fixed or perhaps exposing the start/end dates in the callback again?

**Comment by kaju74** [comment]
If you extract the corresponding methods from the original source code, you'll get:

```
  int _getDaysBefore(DateTime date) {
    return (date.weekday + 7 - getWeekdayNumber(startingDayOfWeek)) % 7;
  }

  int _getDaysAfter(DateTime date) {
    int daysAfter = 7 - ((date.weekday + 8 - getWeekdayNumber(startingDayOfWeek)) % 7);
    return daysAfter == 7 ? 0 : daysAfter;
  }

  DateTime _firstDayOfMonth(DateTime month) {
    return DateTime.utc(month.year, month.month, 1);
  }

  DateTime _lastDayOfMonth(DateTime month) {
    final date = month.month < 12 ? DateTime.utc(month.year, month.month + 1, 1) : DateTime.utc(month.year + 1, 1, 1);
    return date.subtract(const Duration(days: 1));
  }

  DateTimeRange _getVisibleDateRange(DateTime date) {
    late final int daysBefore;
    late final DateTime firstToDisplay;
    late final DateTime lastToDisplay;

    switch (calendarFormat) {
      case CalendarFormat.week:
        daysBefore = _getDaysBefore(date);
        firstToDisplay = date.subtract(Duration(days: daysBefore));
        lastToDisplay = firstToDisplay.add(const Duration(days: 6));
        break;
      case CalendarFormat.twoWeeks:
        daysBefore = _getDaysBefore(date);
        firstToDisplay = date.subtract(Duration(days: daysBefore));
        lastToDisplay = firstToDisplay.add(const Duration(days: 13));
        break;
      default:
        final DateTime first = _firstDayOfMonth(date);
        daysBefore = _getDaysBefore(first);
        firstToDisplay = first.subtract(Duration(days: daysBefore));
        if (sixWeekMonthsEnforced) {
          lastToDisplay = firstToDisplay.add(const Duration(days: 42));
        } else {
          final DateTime last = _lastDayOfMonth(date);
          final int daysAfter = _getDaysAfter(last);
          lastToDisplay = last.add(Duration(days: daysAfter));
        }
    }
    return DateTimeRange(start: firstToDisplay, end: lastToDisplay);
  }
```

So you can implement something like this:

```
  _processPageChanged(DateTime date) {
    DateTimeRange visibleDateRange = _getVisibleDateRange(date);
    Log.debug(
      '_processPageChanged -> Focused date: $date, First Day: ${visibleDateRange.start}, Last Day: ${visibleDateRange.end}',
    );
  }
```

This works as expected but it seems the original code has also problems handling the calendar-format "two weeks". If you've selected a date in the first of the two rows and navigate to the prior or next page, the calculation is correct...but if you've selected a day in the second row, the calculation is shifted one week to the future. 

I didn't find any solution to figure row, on which row the focused day is!?! This way, you could substract one week if placed in the second row...

**Comment by sirajgeelanics** [comment]
> I have the same issue. In previous versions up to 2.3.3 I can get it from: `onVisibleDaysChanged: (first, last, calendarFormat) { }` But this is no longer available in version 3.0.0

But In mine this function was not supporting the 'onVisibleDaysChaged'
